### PR TITLE
Added --changed option to "brew tests" dev cmd

### LINKED ISSUE #13146: Option to only run tests associated with changed files
### Provide a detailed description of the proposed feature

```sh
$ brew tests --changed
```
I was imagining an option like the one above that would only run tests associated with files that have been modified on a feature branch instead of all of the tests.

### What is the motivation for the feature?

It would be much more convenient to do sporadic testing while working on a new feature or bug fix since you wouldn't have to run all of the tests or manually pick the tests you want to run one-by-one. It should also be faster since you are running fewer tests.

### How will the feature be relevant to at least 90% of Homebrew users?

It won't since it's a developer feature but hopefully it will be useful to contributors.

### What alternatives to the feature have been considered?

Keeping things the same.

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This option allows the user to run tests on only the files that have been changed in comparison with the master branch. It slots in between the `brew tests` and `brew tests --only=filename` dev commands.

Resolves #13146.

### CODE REVIEW

File: Library/Homebrew/dev-cmd/tests.rb [hunk 1]
```diff
@@ -124,10 +126,28 @@ def tests
           parallel = false
           ["test/#{test_name}_spec.rb:#{line}"]
         end
+      elsif args.changed?
+        changed_files = Utils.popen_read("git", "diff", "--name-only", "master")
```
Comment (Bo98) [inline Library/Homebrew/dev-cmd/tests.rb:130]: If this works with `git diff`:

```suggestion
        changed_files = Utils.safe_popen_read("git", "diff", "--name-only", "master")
```

(will abort on non-zero exit codes)

**Review by Bo98 — APPROVED** [review]

File: Library/Homebrew/dev-cmd/tests.rb (hunk 1 above)
Comment (apainintheneck) [reply Library/Homebrew/dev-cmd/tests.rb:130]: Yeah, that makes a lot of sense.

Comment (MikeMcQuaid) [reply Library/Homebrew/dev-cmd/tests.rb]: Should be fine as long as we don't use `--exit-code`

File: Library/Homebrew/dev-cmd/tests.rb [hunk 2]
```diff
@@ -26,6 +26,8 @@ def tests_args
                           "official external commands."
       switch "--byebug",
              description: "Enable debugging using byebug."
+      switch "--changed",
+             description: "Only runs tests on files that were changed from the master branch."
       flag   "--only=",
              description: "Run only <test_script>`_spec.rb`. Appending `:`<line_number> will start at a "\
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/tests.rb:32]: Can you mark these as conflicting with `conflicts "--changed", "--only"` below?

File: Library/Homebrew/dev-cmd/tests.rb [hunk 3]
```diff
@@ -124,10 +126,28 @@ def tests
           parallel = false
           ["test/#{test_name}_spec.rb:#{line}"]
         end
+      elsif args.changed?
+        changed_files = Utils.safe_popen_read("git", "diff", "--name-only", "master")
+
+        filestub_regex = %r{Library/Homebrew/([A-Za-z/_-]+).rb}
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/tests.rb:132]:
```suggestion
        filestub_regex = %r{Library/Homebrew/([\w/-]+).rb}
```

File: Library/Homebrew/dev-cmd/tests.rb [hunk 4]
```diff
@@ -124,10 +126,28 @@ def tests
...
+
+        changed_files.scan(filestub_regex)
+                     .map { |filestub| "test/#{filestub.last}_spec.rb" }
+                     .select { |filename| File.exist?(filename) }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/tests.rb:136]:
```suggestion
                     .map { |filestub| Pathname("test/#{filestub.last}_spec.rb") }
                     .select(&:exist?)
```

File: Library/Homebrew/dev-cmd/tests.rb [hunk 5]
```diff
@@ -124,10 +126,28 @@ def tests
...
       else
         Dir.glob("test/**/*_spec.rb")
       end
 
+      if files.empty?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/tests.rb:141]:
```suggestion
      if files.blank?
```

File: Library/Homebrew/dev-cmd/tests.rb [hunk 6]
```diff
@@ -124,10 +126,28 @@ def tests
...
+        raise UsageError, "The --only= argument requires a valid file or folder name!" if args.only
+
+        if args.changed?
+          opoo "Either no files have been changed from the master branch " \
+               "or no tests are associated with the changed files!"
+          return
+        end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/tests.rb:148]:
```suggestion
        if args.only
          raise UsageError, "The --only= argument requires a valid file or folder name!"
        elsif args.changed?
          opoo "Either no files have been changed from the master branch " \
               "or no tests are associated with the changed files!"
          return
        end
```

Also: does the latter want to maybe be a `UsageError` too? Wondering if any other cases need handled, too.

**Review by MikeMcQuaid — COMMENTED** [review]
Great work so far!

File: Library/Homebrew/dev-cmd/tests.rb (hunk 3 above)
Comment (apainintheneck) [reply Library/Homebrew/dev-cmd/tests.rb:132]: Yeah, that's cleaner.

File: Library/Homebrew/dev-cmd/tests.rb (hunk 4 above)
Comment (apainintheneck) [reply Library/Homebrew/dev-cmd/tests.rb:136]: That's a Ruby pattern I didn't know about.

File: Library/Homebrew/dev-cmd/tests.rb (hunk 1 above)
Comment (apainintheneck) [reply Library/Homebrew/dev-cmd/tests.rb]: Ah, okay. I'll change it back.

File: Library/Homebrew/dev-cmd/tests.rb (hunk 6 above)
Comment (apainintheneck) [reply Library/Homebrew/dev-cmd/tests.rb]: I decided to split those two up into a usage error and a warning.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by MikeMcQuaid** [comment]
Thanks again @apainintheneck!

**Pull request merged**